### DISCUSSION TITLE
removable drive access portal

### DISCUSSION THREAD

**Discussion by Be-ing** [flatpak/xdg-desktop-portal#1287] [New Portals]
Some applications present GUIs that treat removable drives separate from other parts of the filesystem. Here is an example from Mixxx, and the [relevant code](https://github.com/mixxxdj/mixxx/blob/d21cb08ce34904192ffd66a855e90da803949664/src/library/browse/browsefeature.cpp#L341-L361):

![image](https://github.com/flatpak/xdg-desktop-portal/assets/9455094/b6d13098-935f-45b4-b9e7-bf934ce4e268)

Currently, this can't work in a Flatpak without specifying
```yaml
  - --filesystem=/media
  - --filesystem=/run/media
```

The file picker portal isn't a solution for this because the application can't determine if there are any devices available to ask the user for permission to use.

See also:
https://github.com/flatpak/flatpak/issues/2713
https://github.com/mixxxdj/mixxx/issues/12860

**Comment by Mikenux** [comment]
If it's about getting files, giving access to the USB storage device makes no sense because it means possibly giving access to files that are irrelevant to the app. For this, as indicated in other issues that you have mentioned, there is the file chooser which allows a finer selection of files. Additionally, I believe the files can be restored by the app, as the app keeps access to the files. So what is the real problem here?

**Reply by Be-ing** [reply]
The problem is that the application can't determine if there even are USB storage devices plugged in, so there's no way to prompt the user for permission to access them.

**Reply by Mikenux** [reply]
I just understood !

What is needed is for a portal to detect when files or folders come from a dynamic location (USB storage, network mount). How to tell this to the app is to see.

**Comment by TingPing** [comment]
So I guess this would be broken down into a few tasks:

- `xdg-desktop-portal` monitors for portable drives being attached (UDisks?)
- A new portal API is added that has a signal applications can listen to, and a method to request access
- The frontends need to present a dialog to the user to allow it. Permissions could be tricky, we both don't want to spam dialogs but also not be too eager to mount things we shouldn't.
  - Probably tie the access to partitions, so if a user reformats a drive access is lost
- Then the directory gets mounted in the sandbox

**Reply by Mikenux** [reply]
> The frontends need to present a dialog to the user to allow it. Permissions could be tricky, we both don't want to spam dialogs but also not be too eager to mount things we shouldn't.

If possible, it would be best to monitor which paths the application has access to, avoiding requesting access.

It would rather be:
- User adds files or folders from a removable drive and the portal is aware of this.
- When the portal knows that removable media is plugged in and mounted, it checks if an application has file or folder permissions.
- The portal can signal that these files and folders are available for access.

**Reply by TingPing** [reply]
> If possible, it would be best to monitor which paths the application has access to, avoiding requesting access.

How is it possible to know a path that doesn't exist yet?

You mean requiring a well known toplevel path?

**Reply by Mikenux** [reply]
Something like this, non-exhaustive:
1. When adding files or folders to the app (file chooser, file manager, drag and drop), the portal knows that they are on a dynamic location (USB storage, network mount).
2. The portal stores information that a given path belongs to a dynamic location and which one (for the respective app).
3. The portal automatically allows the app to subscribe to the signal which will tell it that the files or folders are available. At least, if there is no objection to the fact that an app can automatically know that a file or folder comes from a dynamic location.
4. When the dynamic location is mounted by the system, the monitor is aware of it, checks which apps have permission to access files or folders at that location and sends the signal to the apps.
5. Apps can access the files and folders they have access to.

**Reply by TingPing** [reply]
In my mind a primary use-case for this feature is to plug in a new portable drive and let an app use it.

**Reply by Mikenux** [reply]
Yes, but this is about files. I don't think offering apps the ability to request access to a USB storage device - i.e. "Allow access to [USB_Storage_NAME]?", therefore to access all its files, is a good idea. The fact is that even if we ask for permission, if the user wants to use their USB storage, they will allow access to it, thus giving all the files to the app, without really having the choice.

In contrast, we have the file chooser which allows users to select only the files they want to give to an app: a file, a selection of files, a folder, or all content of a USB storage device (by selecting it in the file chooser).

**Reply by PhL38F** [reply]
Thanks for keeping the topic open and running since it seems very important in practical use cases.

> Yes, but this is about files. I don't think offering apps the ability to request access to a USB storage device - i.e. "Allow access to [USB_Storage_NAME]?", therefore to access all its files, is a good idea. The fact is that even if we ask for permission, if the user wants to use their USB storage, they will allow access to it, thus giving all the files to the app, without really having the choice.

What is the fear behind, here? Bad app writing? Malicious access? User stupidity?

> 
> In contrast, we have the file chooser which allows users to select only the files they want to give to an app: a file, a selection of files, a folder, or all content of a USB storage device (by selecting it in the file chooser).

How can we know that the file chooser is the selected access use chosen by the app writer? What happens if the media is not even first mentioned in the list of devices that one can choose to access ? (I had the case very recently with a flatpack version of Gimp 3.0.4)
NB: LibreOffice that I also run as a flatpack has no issue at all on this kind of storage...

**Reply by Mikenux** [reply]
> What is the fear behind, here? Bad app writing? Malicious access? User stupidity?

Give access to a potentially very large number of files to a malicious app. Giving options to limit access to files as much as possible allows the user to use the app and see if something bad happens to them.

> How can we know that the file chooser is the selected access use chosen by the app writer?

An app writer has only two options for accessing files using portals: drag and drop and file chooser, as they allow the user to limit access to files without restricting them.

> What happens if the media is not even first mentioned in the list of devices that one can choose to access ?

That’s a problem with the UX of the file chooser used by a desktop environment. 

Gimp and LibreOffice have static permissions that allow them to do more than the file chooser portal allows.
